pkg/template/template.go

```diff
@@ -25,32 +27,71 @@ type (
 	// Repository defines all the helper functions to access the generated template specific types and fields.
 	Repository interface {
 		GetTemplateInfo(template string) (Info, bool)
+		DoesBuilderSupportsTemplate(hndlrBuilder adptConfig.HandlerBuilder, tmpl string) (bool, string)
 	}
 	// TypeEvalFn evaluates an expression and returns the ValueType for the expression.
 	TypeEvalFn func(string) (pb.ValueType, error)
 	// InferTypeFn does Type inference from the Constructor.params proto message.
 	InferTypeFn func(proto.Message, TypeEvalFn) (proto.Message, error)
 	// ConfigureTypeFn dispatches the inferred types to handlers
 	ConfigureTypeFn func(types map[string]proto.Message, builder *adptConfig.HandlerBuilder) error
+	// SupportsBuilderFn check if the handlerBuilder supports template.
+	SupportsBuilderFn func(hndlrBuilder adptConfig.HandlerBuilder) bool
 	// Info contains all the information related a template like
 	// Default constructor params, type inference method etc.
 	Info struct {
-		CnstrDefConfig  proto.Message
-		InferTypeFn     InferTypeFn
-		ConfigureTypeFn ConfigureTypeFn
+		CnstrDefConfig    proto.Message
+		InferTypeFn       InferTypeFn
+		ConfigureTypeFn   ConfigureTypeFn
+		SupportsBuilderFn SupportsBuilderFn
+		BuilderName       string
 	}
 	// templateRepo implements Repository
-	templateRepo struct{}
+	templateRepo struct {
+		templateInfos map[string]Info
+
+		allSupportedTmpls  []string
+		tmplToBuilderNames map[string]string
+	}
 )
 
 func (t templateRepo) GetTemplateInfo(template string) (Info, bool) {
-	if v, ok := templateInfos[template]; ok {
+	if v, ok := t.templateInfos[template]; ok {
 		return v, true
 	}
 	return Info{}, false
 }
 
 // NewTemplateRepository returns an implementation of Repository
-func NewTemplateRepository() Repository {
-	return templateRepo{}
+func NewTemplateRepository(templateInfos map[string]Info) Repository {
+	if templateInfos == nil {
+		return templateRepo{
+			templateInfos:      make(map[string]Info),
+			allSupportedTmpls:  make([]string, 0),
+			tmplToBuilderNames: make(map[string]string),
+		}
+	}
+
+	allSupportedTmpls := make([]string, len(templateInfos))
+	tmplToBuilderNames := make(map[string]string)
+
+	for t, v := range templateInfos {
+		allSupportedTmpls = append(allSupportedTmpls, t)
+		tmplToBuilderNames[t] = v.BuilderName
+	}
+	return templateRepo{templateInfos: templateInfos, tmplToBuilderNames: tmplToBuilderNames, allSupportedTmpls: allSupportedTmpls}
+}
+
+func (t templateRepo) DoesBuilderSupportsTemplate(hndlrBuilder adptConfig.HandlerBuilder, tmpl string) (bool, string) {
+	i, ok := t.GetTemplateInfo(tmpl)
+	if !ok {
+		return false, fmt.Sprintf("Supported template %v is not one of the allowed supported templates %v", tmpl, t.allSupportedTmpls)
+	}
+
+	if b := i.SupportsBuilderFn(hndlrBuilder); !b {
```

Comment:
Here and elsewhere with all the function pointers, look at the usage: `if b := i.SupportsBuilderFn(hndlrBuilder); !b { ... }`. It's obvious in context that you're calling a function, you don't need the `Fn` suffix. We put it on types so that it's obvious from the typename that it refers to a function, but there's no need when using an instance of the type.